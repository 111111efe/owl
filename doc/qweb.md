# ðŸ¦‰ QWeb ðŸ¦‰

## Content

- [Overview](#overview)
- [QWeb Specification](#qweb-specification)
    - [Static html nodes](#static-html-nodes)
    - [`t-esc` directive](#t-esc-directive)
    - [`t-raw` directive](#t-raw-directive)
    - [`t-set` directive](#t-set-directive)
    - [`t-if` directive](#t-if-directive)
    - [Expression evaluation](#expression-evaluation)
    - [`t-att` directive (dynamic attributes)](#t-att-directive-dynamic-attributes)
- [JS/OWL Specific Extensions](#jsowl-specific-extensions)
    - [`t-on` directive](#t-on-directive)
    - [Component: `t-widget`, `t-props`](#component-t-widget-t-props)
    - [`t-ref` directive](#t-ref-directive)
    - [`t-key` directive](#t-key-directive)
    - [Debugging (`t-debug` and `t-log`)](#debugging-t-debug-and-t-log)
    - [White spaces](#white-spaces)
    - [Root nodes](#root-nodes)

## Overview

[QWeb](https://www.odoo.com/documentation/12.0/reference/qweb.html) is the primary templating engine used by Odoo. It is based on the XML format, and used
mostly to generate html. In OWL, QWeb templates are compiled into functions that
generate a virtual dom representation of the html.

Template directives are specified as XML attributes prefixed with `t-`, for instance `t-if` for conditionals, with elements and other attributes being rendered directly.

To avoid element rendering, a placeholder element `<t>` is also available, which executes its directive but doesnâ€™t generate any output in and of itself.

The QWeb implementation in the OWL project is slightly different. It compiles
templates into functions that output a virtual DOM instead of a string. This is
necessary for the component system. In addition, it has a few extra directives
(see [OWL Specific Extensions](#owlspecificextensions))


## QWeb Specification


### Static html nodes

Normal, regular html nodes are rendered into themselves:

```xml
  <div>hello</div> <!â€“â€“ rendered as itself â€“â€“>
```

### `t-esc` directive

The `t-esc` directive is necessary whenever you want to add a dynamic text
expression in a template. The text is escaped to avoid security issues.

```xml
<p><t t-esc="value"/></p>
```

rendered with the value `value` set to `42` in the rendering context yields:

```html
<p>42</p>
```

### `t-raw` directive

The `t-raw` directive is almost the same as `t-esc`, but without the escaping.
This is mostly useful to inject a raw html string somewhere. Obviously, this
is unsafe to do in general, and should only be used for strings known to be safe.

```xml
<p><t t-esc="value"/></p>
```

rendered with the value `value` set to `<span>foo</span>` in the rendering context yields:

```html
<p><span>foo</span></p>
```

### `t-set` directive

QWeb allows creating variables from within the template, to memoize a computation (to use it multiple times), give a piece of data a clearer name, ...

This is done via the `t-set` directive, which takes the name of the variable to create. The value to set can be provided in two ways:

1. a `t-value` attribute containing an expression, and the result of its
   evaluation will be set:

   ```xml
   <t t-set="foo" t-value="2 + 1"/>
   <t t-esc="foo"/>
   ```

   will print `3`. Note that the evaluation is done at rendering time, not at
   compilte time.

2. if there is no `t-value` attribute, the nodeâ€™s body is saved and its value is
   set as the variableâ€™s value:

   ```xml
   <t t-set="foo">
       <li>ok</li>
   </t>
   <t t-esc="foo"/>
   ```

   will generate `&lt;li&gt;ok&lt;/li&gt;` (the content is escaped as we used the `t-esc` directive)

The `t-set` directive acts like a regular variable in most programming language.
It is lexically scoped (inner nodes are sub scopes), can be shadowed, ...

### `t-if` directive

The `t-if` directive is useful to conditionally render something. It evaluates
the expression given as attribute value, and then acts accordingly.

```xml
<div>
    <t t-if="condition">
        <p>ok</p>
    </t>
</div>
```

The element is rendered if the condition (evaluated with the current rendering
context) is true:

```xml
<div>
    <p>ok</p>
</div>
```

but if the condition is false it is removed from the result:

```xml
<div>
</div>
```

The conditional rendering applies to the bearer of the directive, which does not
have to be `<t>`:

```xml
<div>
    <p t-if="condition">ok</p>
</div>
```

will give the same results as the previous example.

Extra conditional branching directives `t-elif` and `t-else` are also available:

```xml
<div>
    <p t-if="user.birthday == today()">Happy bithday!</p>
    <p t-elif="user.login == 'root'">Welcome master!</p>
    <p t-else="">Welcome!</p>
</div>
```

### Expression evaluation

It is useful to explain the various rules that applies on QWeb expressions. These
expressions are strings that will be converted to a javascript expression at
compile time.

1. it should be a simple expression which returns a value. It cannot be a statement.

   ```xml
   <div><p t-if="1 + 2 === 3">ok</p></div>
   ```

   is valid, but the following is not valid:

   ```xml
   <div><p t-if="console.log(1)">NOT valid</p></div>
   ```

2. it can use anything in the rendering context:

   ```xml
   <p t-if="user.birthday == today()">Happy bithday!</p>
   ```

   is valid, and will read the `user` object from the context, and call the
   `today` function.

3. it can use a few special operators to avoid using symbols such as `<`, `>`,
   `&` or `|`. This is useful to make sure that we still write valid XML.

   | Word  | will be replaced by |
   | ----- | ------------------- |
   | `and` | `&&`                |
   | `or`  | `\|\|`                |
   | `gt`  | `>`                 |
   | `gte` | `>=`                |
   | `lt`  | `<`                 |
   | `lte` | `<=`                |

   So, one can write this:

   ```xml
   <div><p t-if="10 + 2 gt 5">ok</p></div>
   ```

### `t-att` directive (dynamic attributes)

One can use the `t-att-` directive to add dynamic attributes. Its main use is to
evaluate an expression (at rendering time) and bind an attribute to its result:

For example, if we have `id` set to 32 in the rendering context,

```xml
<div t-att-data-action-id="id"/>  <!-- result: <div data-action-id="32"></div> -->
```

If an expression evaluates to a falsy value, it will not be set at all:

```xml
<div t-att-foo="false"/>  <!-- result: <div></div> -->
```

## JS/OWL Specific Extensions

### `t-on` directive

### Component: `t-widget`, `t-props`

### `t-ref` directive

### `t-key` directive

Even though Owl tries to be as declarative as possible, some DOM state is still
locked inside the DOM: for example, the scrolling state, the current user selection,
the focused element or the state of an input. This is why we use a virtual dom
algorithm to keep the actual DOM node as much as possible.  However, this is
sometimes not enough, and we need to help Owl decide if an element is actually
the same, or is different. The `t-key` directive is used to give an identity to an element.

There are three main use cases:

- *elements in a list*:
  ```xml
    <span t-foreach="todos" t-as="todo" t-key="todo.id">
        <t t-esc="todo.text"/>
    </span>
  ```

- *`t-if`/`t-else`*

- *animations*: give a different identity to a component.  Ex: thread id with
animations on add/remove message.

### Debugging (`t-debug` and `t-log`)

The javascript QWeb implementation provides two useful debugging directives:

`t-debug` adds a debugger statement during template rendering:

```xml
<t t-if="a_test">
    <t t-debug="">
</t>
````

will stop execution if the browser dev tools are open.

`t-log` takes an expression parameter, evaluates the expression during rendering and logs its result with console.log:

```xml
<t t-set="foo" t-value="42"/>
<t t-log="foo"/>
```

will print 42 to the console


### White spaces

White spaces in a templates are handled in a special way:

- consecutive whitespaces are always condensed to a single whitespace
- if a whitespace-only text node contains a linebreak, it is ignored
- the previous rules do not apply if we are in a `<pre>` tag

### Root nodes

For many reasons, Owl QWeb templates should have a single root node.  More
precisely, the result of a template rendering should have a single root node:

```xml
<!â€“â€“ not ok: two root nodes â€“â€“>
<t>
    <div>foo</div>
    <div>bar</div>
</t>

<!â€“â€“ ok: result has one single root node â€“â€“>
<t>
    <div t-if="someCondition">foo</div>
    <span t-else="1">bar</span>
</t>
```

Extra root nodes will actually be ignored (even though they will be rendered
in memory).

Note: this does not apply to subtemplates (see the `t-call` directive).  In that
case, they will be inlined in the main template, and can actually have many
root nodes.