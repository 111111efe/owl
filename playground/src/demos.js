export const HELLO_WORLD = `class HelloWorld extends owl.core.Component {
  inlineTemplate = \`<div>Hello <t t-esc="props.name"/></div>\`;
}

const env = {
    qweb: new owl.core.QWeb()
};

const hello = new HelloWorld(env, { name: "World" });
hello.mount(document.body);
`;
