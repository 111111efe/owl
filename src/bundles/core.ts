import QWeb from "../core/qweb_vdom";
import * as utils from "../core/utils";
import Widget from "../core/widget";


(<any>window).core = {
    QWeb,
    utils,
    Widget
}