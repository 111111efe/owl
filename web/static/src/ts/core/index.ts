export { QWeb } from "./qweb_vdom";
export { EventBus } from "./event_bus";
export { Component } from "./component";
