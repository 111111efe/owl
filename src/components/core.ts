import { qweb } from "../qweb/qweb";
import { QWebTemplate } from "../qweb/types";
import { VRootNode } from "../vdom/types";
import { buildTree, patch } from "../vdom/vdom";
import { Component } from "./component";
import { Fiber } from "./fiber";
import { scheduler } from "./scheduler";

// -----------------------------------------------------------------------------
// Types
// -----------------------------------------------------------------------------

export interface FComponent<T = any, Env = any> {
  template: string;
  components?: { [key: string]: FComponent<any> | CComponent };
  name?: string;
  setup?: (props: any, env: Env) => T | Promise<T> | void;
}

export type CComponent = typeof Component;

export type CInstance<Props = any, Env = any> = Component<Props, Env>;

export type FInstance<T = {}, Env = any> = T & {
  __owl__: OwlElement<T>;
  props: any;
  env: Env;
};

export interface OwlElement<T = any> {
  fiber: Fiber | null;
  name: string;
  components: { [key: string]: FComponent<T> | CComponent };
  isMounted: boolean;
  instance: FInstance<T> | CInstance | null;
  vnode: VRootNode | null;
  qweb: QWebTemplate;
  isReady: Promise<void> | null;
  children: { [key: string]: OwlElement };
}

// -----------------------------------------------------------------------------
// core object
// -----------------------------------------------------------------------------
export const core = {
  current: null as null | OwlElement,
};

// -----------------------------------------------------------------------------
// Mount
// -----------------------------------------------------------------------------
export type MountTarget = HTMLElement | DocumentFragment;

export interface MountConfig<Env = any> {
  props?: Object;
  env?: Env;
}

export function mount<C extends CComponent>(
  Comp: C,
  target: MountTarget,
  config?: MountConfig
): Promise<InstanceType<C>>;
export function mount(
  comp: CInstance,
  target: MountTarget,
  config?: MountConfig
): Promise<CInstance>;
export function mount<T, Env>(
  Comp: FComponent<T>,
  target: MountTarget,
  config?: MountConfig<Env>
): Promise<FInstance<T extends {} ? T : {}, Env>>;
export function mount<T, Env>(
  comp: FInstance,
  target: MountTarget,
  config?: MountConfig<Env>
): Promise<FInstance<T extends {} ? T : {}, Env>>;
export async function mount(
  info: any,
  target: MountTarget,
  config: MountConfig = {}
): Promise<CInstance | FInstance> {
  if (!(target instanceof HTMLElement || target instanceof DocumentFragment)) {
    const name =
      info instanceof Component ? info.constructor.name : (info as any).name || "No Name";
    let message = `Component '${name}' cannot be mounted: the target is not a valid DOM node.`;
    message += `\nMaybe the DOM is not ready yet? (in that case, you can use owl.utils.whenReady)`;
    throw new Error(message);
  }
  if (info instanceof Component) {
    // this is a CInstance
    const elem = info.__owl__!;
    return scheduler.addFiber(elem.fiber!).then(() => {
      target.appendChild(info.el!);
      return info;
    });
  }
  let element: OwlElement;
  let props = config.props || {};
  let env = config.env || {};
  if (info.prototype instanceof Component || info === Component) {
    // this is a CComponent (Class Component)
    element = createCComponent(info, props, env);
  } else if ("__owl__" in info) {
    // this is a CInstance or a FInstance
    element = info.__owl__;
  } else {
    // this is a FComponent
    element = createFComponent(info, props, env);
  }
  return scheduler.addFiber(element.fiber!).then(() => {
    const fragment = document.createDocumentFragment();
    buildTree(element.vnode!, (fragment as any) as HTMLElement);
    target.appendChild(fragment);
    if (document.body.contains(target)) {
      mountTree(element);
    }
    return element.instance!;
  });
}

function mountTree(element: OwlElement) {
  for (let k in element.children) {
    mountTree(element.children[k]);
  }
  element.isMounted = true;
  if (element.instance.mounted) {
    element.instance.mounted();
  }
}

// -----------------------------------------------------------------------------
// Component creation
// -----------------------------------------------------------------------------

export function createFComponent(fn: FComponent, props: any, env: any): OwlElement {
  let template: string = fn.template;
  const name = fn.name || "Anonymous Function Component";
  if (!template) {
    throw new Error(`Component "${name}" does not have a template defined!`);
  }
  const qtemplate = qweb.getTemplate(template);
  const elem: OwlElement = {
    fiber: null,
    name,
    instance: null,
    isMounted: false,
    vnode: null,
    components: fn.components || {},
    qweb: qtemplate,
    isReady: null,
    children: {},
  };
  core.current = elem;

  if (fn.setup) {
    const result = fn.setup(props, env);
    const prom = result instanceof Promise ? result : Promise.resolve(result);
    elem.isReady = prom.then((result) => {
      let instance = result ? Object.create(result) : {};
      instance.props = props;
      instance.__owl__ = elem;
      instance.env = env;
      elem.instance = instance;
    });
  } else {
    const instance: FInstance = { props, env, __owl__: elem };
    elem.instance = instance;
    elem.isReady = Promise.resolve();
  }

  startComponent(elem);
  return elem;
}

export function createCComponent<C extends typeof Component>(
  C: C,
  props: any,
  env: any
): OwlElement {
  let template: string = C.template;
  if (!template) {
    throw new Error(`Component "${C.name}" does not have a template defined!`);
  }
  const qtemplate = qweb.getTemplate(template);
  const elem: OwlElement = {
    fiber: null,
    name: C.name,
    instance: null,
    isMounted: false,
    vnode: null,
    components: C.components,
    qweb: qtemplate,
    isReady: null,
    children: {},
  };
  core.current = elem;
  const component = new C(props, env) as InstanceType<C>;
  component.__owl__ = elem;
  elem.instance = component;
  elem.isReady = component.willStart();
  startComponent(elem);
  return elem;
}

// -----------------------------------------------------------------------------
// starting Components
// -----------------------------------------------------------------------------

async function startComponent(elem: OwlElement) {
  const vnode = elem.qweb.createRoot();
  vnode.hooks.create = (el) => (elem.instance.el = el);
  const fiber = new Fiber(elem, null, vnode);
  fiber.root.counter++;
  elem.fiber = fiber;
  elem.vnode = vnode;
  await elem.isReady!;
  elem.qweb.render(vnode, elem.instance, elem);
  fiber.root.counter--;
}

export function renderToFiber(elem: OwlElement, parentFiber: Fiber | null): Fiber {
  const newRoot = elem.qweb.createRoot();
  const fiber = new Fiber(elem, parentFiber, newRoot);
  elem.fiber = fiber;
  fiber.root.counter++;
  new Promise((resolve) => {
    elem.qweb.render(newRoot, elem.instance!, elem);
    fiber.root.counter--;
    resolve();
  });
  return fiber;
}
// -----------------------------------------------------------------------------
// render
// -----------------------------------------------------------------------------

export function render(elem: OwlElement): Promise<void> {
  let fiber = renderToFiber(elem, null);
  return scheduler.addFiber(fiber).then(() => {
    patch(elem.vnode!, fiber.vnode!);
    while (fiber.next) {
      fiber = fiber.next;
      patch(fiber.elem.vnode!, fiber.vnode!);
    }
  });
}
