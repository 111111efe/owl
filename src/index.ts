import {} from "./components/component";
import { mount, Component } from "./components/index";
import { qweb } from "./qweb/qweb";
import { useState } from "./hooks";
import { xml } from "./tags";

const renderToString = qweb.renderToString;

export { mount, renderToString, Component, xml, useState };
