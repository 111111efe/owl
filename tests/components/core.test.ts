import { Component, mount, xml, useState } from "../../src/index";
import { makeTestFixture, nextTick } from "../helpers";
import { FComponent } from "../../src/components/core";
import { qweb } from "../../src/qweb/qweb";

//------------------------------------------------------------------------------
// Setup and helpers
//------------------------------------------------------------------------------

let fixture: HTMLElement;

beforeEach(() => {
  fixture = makeTestFixture();
});

afterEach(() => {
  fixture.remove();
});

//------------------------------------------------------------------------------
// Tests
//------------------------------------------------------------------------------

describe("basic component properties", () => {
  describe("mounting components", () => {
    test("can mount a simple functional component", async () => {
      const Test = {
        template: xml`<div>simple vnode</div>`,
      };

      await mount(Test, fixture);
      expect(fixture.innerHTML).toBe("<div>simple vnode</div>");
      expect(qweb.compiledTemplates[Test.template].fn).toMatchSnapshot();
    });

    test("can mount a simple class component", async () => {
      class Test extends Component {
        static template = xml`<div>simple vnode</div>`;
      }

      const test = await mount(Test, fixture);
      expect(fixture.innerHTML).toBe("<div>simple vnode</div>");
      expect(test).toBeInstanceOf(Component);

      const div = fixture.querySelector("div");
      expect(div).toBeDefined();
      expect(test.el).toBe(div);
    });

    test("return value of mount contains the result of setup", async () => {
      const obj = {};
      const Test = {
        template: xml`<div>simple vnode</div>`,
        setup() {
          return obj;
        },
      };

      const test = await mount(Test, fixture);
      test.__owl__;
      expect((test as any).__proto__).toBe(obj);
      expect(fixture.innerHTML).toBe("<div>simple vnode</div>");
    });

    // ---------------------------------------------------------------------------
    // only a text node
    // ---------------------------------------------------------------------------
    test("functional component, just a text node", async () => {
      const Test = {
        template: xml`simple text node`,
      };
      await mount(Test, fixture);
      expect(fixture.innerHTML).toBe("simple text node");
    });

    test("class component, text node", async () => {
      class Test extends Component {
        static template = xml`simple text node`;
      }

      await mount(Test, fixture);
      expect(fixture.innerHTML).toBe("simple text node");
    });

    // ---------------------------------------------------------------------------
    // multiroot content
    // ---------------------------------------------------------------------------

    test("simple functional component, multiroot", async () => {
      const Test = {
        template: xml`<div>a</div><div>b</div>`,
      };

      await mount(Test, fixture);
      expect(fixture.innerHTML).toBe("<div>a</div><div>b</div>");
    });

    test("simple class component, multiroot", async () => {
      class Test extends Component {
        static template = xml`<div>a</div><div>b</div>`;
      }

      await mount(Test, fixture);
      expect(fixture.innerHTML).toBe("<div>a</div><div>b</div>");
    });

    // ---------------------------------------------------------------------------
    // simple mounting operation, dynamic content
    // ---------------------------------------------------------------------------

    test("functional component with dynamic content", async () => {
      const Test = {
        template: xml`<div>Hello <t t-esc="name"/></div>`,
        setup() {
          return { name: "Alex" };
        },
      };

      await mount(Test, fixture);
      expect(fixture.innerHTML).toBe("<div>Hello Alex</div>");
    });

    test("class component with dynamic content", async () => {
      class Test extends Component {
        static template = xml`<div>Hello <t t-esc="name"/></div>`;
        name = "Alex";
      }

      await mount(Test, fixture);
      expect(fixture.innerHTML).toBe("<div>Hello Alex</div>");
    });

    test("mounting  Component without template throw errors (*)", async () => {
      let e: Error | null = null;
      try {
        await mount(Component, fixture);
      } catch (error) {
        e = error;
      }
      expect(e).toBeDefined();
      expect(e!.message).toBe('Component "Component" does not have a template defined!');
    });

    test("mounting a function Component without template throw errors (*)", async () => {
      let e: Error | null = null;
      const Test = {};
      try {
        await mount(Test as any, fixture);
      } catch (error) {
        e = error;
      }
      expect(e).toBeDefined();
      expect(e!.message).toBe(
        'Component "Anonymous Function Component" does not have a template defined!'
      );
    });

    test("mounting a Class Component without template throw error (*)", async () => {
      class D extends Component {}
      let e: Error | null = null;
      try {
        await mount(D, fixture);
      } catch (error) {
        e = error;
      }
      expect(e).toBeDefined();
      expect(e!.message).toBe('Component "D" does not have a template defined!');
    });

    test("mounting a Class Component  with a static template key not registered (*)", async () => {
      class D extends Component {
        static template = `<div>abc</div>`;
      }

      let e: Error | null = null;
      try {
        await mount(D, fixture);
      } catch (error) {
        e = error;
      }
      expect(e).toBeDefined();
      expect(e!.message).toBe(
        'Cannot find template with name "<div>abc</div>". Maybe you should register it with "xml" helper.'
      );
    });

    test("props is set on root class Components", async () => {
      class Test extends Component {
        static template = xml`<div></div>`;
      }
      const test = await mount(Test, fixture);
      expect(test.props).toEqual({});
    });

    test("functional component: props is received in setup function", async () => {
      expect.assertions(1);
      const Test = {
        template: xml`<div></div>`,
        setup(props: any) {
          expect(props).toEqual({});
        },
      };

      await mount(Test, fixture);
    });

    test("env is set on root class component (*)", async () => {
      class Test extends Component {
        static template = xml`<div></div>`;
      }
      const test = await mount(Test, fixture);
      expect(test.env).toEqual({});
    });

    test("env is set on root fcomponent (*)", async () => {
      expect.assertions(1);

      const Test: FComponent = {
        template: xml`<div></div>`,
        setup(props, env) {
          expect(env).toEqual({});
        },
      };
      await mount(Test, fixture);
    });

    test("configured env is set on root class component (*)", async () => {
      class Test extends Component {
        static template = xml`<div></div>`;
      }
      const env = { a: 1 };
      const test = await mount(Test, fixture, { env });
      expect(test.env).toEqual(env);
    });

    test("configured env is set on root fcomponent (*)", async () => {
      expect.assertions(2);
      const env = { a: 1 };

      const Test = {
        template: xml`<div></div>`,
        setup(props: any, env: any) {
          expect(env).toBe(env);
        },
      };
      const test = await mount(Test, fixture, { env });
      expect(test.env).toBe(env);
    });

    test("can give props to class component with mount method", async () => {
      class Test extends Component {
        static template = xml`<div></div>`;
      }
      const p = { a: 1 };
      const test = await mount(Test, fixture, { props: p });
      expect(test.props).toBe(p);
    });

    test("can give props to fcomponent with mount method", async () => {
      expect.assertions(1);
      const p = { a: 1 };

      const Test = {
        template: xml`<div></div>`,
        setup(props: any) {
          expect(props).toBe(p);
        },
      };

      await mount(Test, fixture, { props: p });
    });

    test("class component has no el after creation", async () => {
      expect.assertions(2);
      class Test extends Component {
        static template = xml`<div></div>`;
        constructor(props: any, env: any) {
          super(props, env);
          expect(this.el).toBeNull();
        }
      }
      const test = await mount(Test, fixture);
      expect(test.el).not.toBe(null);
    });

    test("class component can be mounted in a div", async () => {
      class SomeWidget extends Component {
        static template = xml`<div>content</div>`;
      }
      await mount(SomeWidget, fixture);
      expect(fixture.innerHTML).toBe("<div>content</div>");
    });

    test("fcomponent can be mounted in a div", async () => {
      const Test = {
        template: xml`<div>content</div>`,
      };
      await mount(Test, fixture);
      expect(fixture.innerHTML).toBe("<div>content</div>");
    });

    test("class component can be mounted on a documentFragment, then mounted in a div", async () => {
      class Test extends Component {
        static template = xml`<div>content</div>`;
      }
      const fragment = document.createDocumentFragment();
      const test = await mount(Test, fragment);
      expect(fixture.innerHTML).toBe("");
      const result = await mount(test, fixture);
      expect(fixture.innerHTML).toBe("<div>content</div>");
      expect(result).toBe(test);
    });

    test("fcomponent can be mounted on a documentFragment, then mounted in a div", async () => {
      const Test = {
        template: xml`<div>content</div>`,
      };
      const fragment = document.createDocumentFragment();
      const test = await mount(Test, fragment);
      expect(fixture.innerHTML).toBe("");
      await mount(test, fixture);
      expect(fixture.innerHTML).toBe("<div>content</div>");
    });

    test("class component: display a nice error message if mounted on a non existing node", async () => {
      class Test extends Component {
        static template = xml`<div>content</div>`;
      }
      let error;
      try {
        await mount(Test, null as any);
      } catch (e) {
        error = e;
      }
      expect(error).toBeDefined();
      expect(error.message).toBe(
        "Component 'Test' cannot be mounted: the target is not a valid DOM node.\nMaybe the DOM is not ready yet? (in that case, you can use owl.utils.whenReady)"
      );
    });

    test("fcomponent: display a nice message if mounted on a non existing node", async () => {
      const Test = {
        template: xml`<div>content</div>`,
      };
      let error;
      try {
        await mount(Test, null as any);
      } catch (e) {
        error = e;
      }
      expect(error).toBeDefined();
      expect(error.message).toBe(
        "Component 'No Name' cannot be mounted: the target is not a valid DOM node.\nMaybe the DOM is not ready yet? (in that case, you can use owl.utils.whenReady)"
      );
    });

    test("class component: display a nice message if mounted on an invalid node (*)", async () => {
      class Test extends Component {
        static template = xml`<div>content</div>`;
      }
      let error;
      try {
        await mount(Test, {} as any);
      } catch (e) {
        error = e;
      }
      expect(error).toBeDefined();
      expect(error.message).toBe(
        "Component 'Test' cannot be mounted: the target is not a valid DOM node.\nMaybe the DOM is not ready yet? (in that case, you can use owl.utils.whenReady)"
      );
    });

    test("fcomponent: display a nice message if mounted on an invalid node (*)", async () => {
      const Test = {
        template: xml`<div>content</div>`,
      };
      let error;
      try {
        await mount(Test, {} as any);
      } catch (e) {
        error = e;
      }
      expect(error).toBeDefined();
      expect(error.message).toBe(
        "Component 'No Name' cannot be mounted: the target is not a valid DOM node.\nMaybe the DOM is not ready yet? (in that case, you can use owl.utils.whenReady)"
      );
    });

    test("ccomponent: can be clicked on and updated", async () => {
      class Counter extends Component {
        static template = xml`
          <div>
            <t t-esc="state.counter"/>
            <button t-on-click="state.counter++">Inc</button>
          </div>`;
        state = useState({
          counter: 0,
        });
      }

      await mount(Counter, fixture);

      expect(fixture.innerHTML).toBe("<div>0<button>Inc</button></div>");
      fixture.querySelector("button")!.click();
      await nextTick();
      expect(fixture.innerHTML).toBe("<div>1<button>Inc</button></div>");
    });

    test("fcomponent: can be clicked on and updated", async () => {
      const Counter = {
        template: xml`
          <div>
            <t t-esc="state.counter"/><button t-on-click="state.counter++">Inc</button>
          </div>`,
        setup() {
          const state = useState({ counter: 0 });
          return { state };
        },
      };

      await mount(Counter, fixture);
      expect(fixture.innerHTML).toBe("<div>0<button>Inc</button></div>");
      fixture.querySelector("button")!.click();
      await nextTick();
      expect(fixture.innerHTML).toBe("<div>1<button>Inc</button></div>");
    });

    test("ccomponent: can handle empty props", async () => {
      class Child extends Component {
        static template = xml`<span><t t-esc="props.val"/></span>`;
      }
      class Parent extends Component {
        static template = xml`<div><Child val=""/></div>`;
        static components = { Child };
      }

      await mount(Parent, fixture);
      expect(qweb.compiledTemplates[Parent.template].fn.toString()).toMatchSnapshot();
      expect(fixture.innerHTML).toBe("<div><span></span></div>");
    });

    test("fcomponent: can handle empty props", async () => {
      const Child = {
        template: xml`<span><t t-esc="props.val"/></span>`,
      };
      const Parent = {
        template: xml`<div><Child val=""/></div>`,
        components: { Child },
      };

      await mount(Parent, fixture);
      expect(qweb.compiledTemplates[Parent.template].fn.toString()).toMatchSnapshot();
      expect(fixture.innerHTML).toBe("<div><span></span></div>");
    });

    test("ccomponent: can access env in template (*)", async () => {
      class Test extends Component {
        static template = xml`<span><t t-esc="env.val"/></span>`;
      }

      await mount(Test, fixture, { env: { val: 3 } });
      expect(fixture.innerHTML).toBe("<span>3</span>");
    });

    test("fcomponent: can access env in template (*)", async () => {
      const Test = {
        template: xml`<span><t t-esc="env.val"/></span>`,
      };

      await mount(Test, fixture, { env: { val: 3 } });
      expect(fixture.innerHTML).toBe("<span>3</span>");
    });

    test("ccomponents are flagged mounted if in dom (*)", async () => {
      class Test extends Component {
        static template = xml`<div></div>`;
      }

      const target = document.createElement("div");
      const test1 = await mount(Test, target);
      expect(test1.__owl__!.isMounted).toBe(false);
      const test2 = await mount(Test, fixture);
      expect(test2.__owl__!.isMounted).toBe(true);
    });

    test("fcomponents are flagged mounted if in dom (*)", async () => {
      const Test = {
        template: xml`<div></div>`,
      };

      const target = document.createElement("div");
      const test1 = await mount(Test, target);
      expect(test1.__owl__.isMounted).toBe(false);
      const test2 = await mount(Test, fixture);
      expect(test2.__owl__.isMounted).toBe(true);
    });

    test("ccomponent: cannot be clicked on and updated if not in DOM", async () => {
      class Counter extends Component {
        static template = xml`
          <div>
            <t t-esc="state.counter"/>
            <button t-on-click="increment">Inc</button>
          </div>`;
        state = useState({
          counter: 0,
        });
        increment() {
          this.state.counter++;
        }
      }

      const target = document.createElement("div");
      const counter = await mount(Counter, target);
      expect(target.innerHTML).toBe("<div>0<button>Inc</button></div>");
      const button = target.getElementsByTagName("button")[0];
      button.click();
      await nextTick();
      expect(target.innerHTML).toBe("<div>0<button>Inc</button></div>");
      expect(counter.state.counter).toBe(0);
    });

    test("fcomponent: cannot be clicked on and updated if not in DOM", async () => {
      const Counter = {
        template: xml`
          <div>
            <t t-esc="state.counter"/>
            <button t-on-click="state.counter++">Inc</button>
          </div>`,
        setup() {
          return { state: useState({ counter: 0 }) };
        },
      };

      const target = document.createElement("div");
      const counter = await mount(Counter, target);
      expect(target.innerHTML).toBe("<div>0<button>Inc</button></div>");
      const button = target.querySelector("button")!;
      button.click();
      await nextTick();
      expect(target.innerHTML).toBe("<div>0<button>Inc</button></div>");
      expect(counter.state.counter).toBe(0);
    });

    test("ccomponent with style and classname", async () => {
      class Test extends Component {
        static template = xml`
          <div style="font-weight:bold;" class="some-class">world</div>
        `;
      }
      await mount(Test, fixture);
      expect(fixture.innerHTML).toBe(
        `<div style="font-weight:bold;" class="some-class">world</div>`
      );
    });

    test("fcomponent with style and classname", async () => {
      const Test = {
        template: xml`
          <div style="font-weight:bold;" class="some-class">world</div>`,
      };
      await mount(Test, fixture);
      expect(fixture.innerHTML).toBe(
        `<div style="font-weight:bold;" class="some-class">world</div>`
      );
    });

    test("ccomponent: keeps a reference to env", async () => {
      class Test extends Component {
        static template = xml`<div></div>`;
      }
      const env = { a: 3 };
      const test = await mount(Test, fixture, { env });
      expect(test.env).toBe(env);
    });
  });

  describe("rerendering components", () => {
    test("ccomponent: render method wait until rendering is done", async () => {
      class Test extends Component {
        static template = xml`<div><t t-esc="state.drinks"/></div>`;
        state = { drinks: 1 };
      }
      const test = await mount(Test, fixture);
      expect(fixture.innerHTML).toBe("<div>1</div>");

      test.state.drinks = 2;

      const renderPromise = test.render();
      expect(fixture.innerHTML).toBe("<div>1</div>");
      await renderPromise;
      expect(fixture.innerHTML).toBe("<div>2</div>");
    });

    test("ccomponent: do not remove previously rendered dom if not necessary", async () => {
      class Test extends Component {
        static template = xml`<div/>`;
      }
      const test = await mount(Test, fixture);
      expect(fixture.innerHTML).toBe(`<div></div>`);

      test.el!.appendChild(document.createElement("span"));
      expect(fixture.innerHTML).toBe(`<div><span></span></div>`);
      await test.render();
      expect(fixture.innerHTML).toBe(`<div><span></span></div>`);
    });
  });
});
