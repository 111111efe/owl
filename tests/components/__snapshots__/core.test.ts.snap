// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`basic component properties mounting components ccomponent: can handle empty props 1`] = `
"function anonymous(tree, ctx, metadata
) {
    // Template: \`__template__30\`
    const vn1 = tree.child = {type: 2, tag: \\"div\\", children: [], key: 1};
    let _2 = {val:undefined};
    vn1.children.push(this.makeComponent(metadata, \\"Child\\", ctx, _2));
}"
`;

exports[`basic component properties mounting components fcomponent: can handle empty props 1`] = `
"function anonymous(tree, ctx, metadata
) {
    // Template: \`__template__32\`
    const vn1 = tree.child = {type: 2, tag: \\"div\\", children: [], key: 1};
    let _2 = {val:undefined};
    vn1.children.push(this.makeComponent(metadata, \\"Child\\", ctx, _2));
}"
`;
